execution_engine_testing/tests/src/test/contract_api/transfer_cached.rs

```diff
@@ -0,0 +1,191 @@
+
+    builder.run_genesis(&DEFAULT_RUN_GENESIS_REQUEST);
+
+    let pre_state_hash = builder.get_post_state_hash();
+
+    // Default account to account 1
+    let mut exec_builder = ExecuteRequestBuilder::new();
+    // We must first transfer the amount account 1 will transfer to account 2, along with the fee
+    // account 1 will need to pay for that transfer.
+    exec_builder = exec_builder.push_deploy(transfer(
+        *DEFAULT_ACCOUNT_ADDR,
+        runtime_args! {
+           ARG_TARGET => *ACCOUNT_1_ADDR,
+           ARG_AMOUNT => *TRANSFER_AMOUNT + TRANSFER_COST,
+           ARG_ID => <Option<u64>>::None,
+        },
+    ));
+    builder.scratch_exec(exec_builder.build()).expect_success();
+
+    let mut exec_builder = ExecuteRequestBuilder::new();
+    exec_builder = exec_builder.push_deploy(transfer(
+        *ACCOUNT_1_ADDR,
+        runtime_args! {
+            ARG_TARGET => *ACCOUNT_2_ADDR,
+            ARG_AMOUNT => *TRANSFER_AMOUNT,
+            ARG_ID => <Option<u64>>::None,
+        },
+    ));
+
+    builder.scratch_exec(exec_builder.build()).expect_success();
+
+    // Double spend test for account 1
+    let mut exec_builder = ExecuteRequestBuilder::new();
+    exec_builder = exec_builder.push_deploy(transfer(
+        *ACCOUNT_1_ADDR,
+        runtime_args! {
+            ARG_TARGET => *ACCOUNT_2_ADDR,
+            ARG_AMOUNT => *TRANSFER_AMOUNT,
+            ARG_ID => <Option<u64>>::None,
+        },
+    ));
+
+    builder.scratch_exec(exec_builder.build()).expect_failure();
+
+    builder.write_scratch_to_lmdb();
+    builder.flush_environment();
+
+    assert_ne!(
+        pre_state_hash,
+        builder.get_post_state_hash(),
+        "post state hash didn't change..."
+    );
+
+    let default_account = builder
+        .get_account(*DEFAULT_ACCOUNT_ADDR)
+        .expect("should get default account");
+
+    let account1 = builder
+        .get_account(*ACCOUNT_1_ADDR)
+        .expect("should get account 1");
+
+    let account2 = builder
+        .get_account(*ACCOUNT_2_ADDR)
+        .expect("should get account 2");
+
+    let default_account_balance = builder.get_purse_balance(default_account.main_purse());
+    let default_expected_balance = U512::from(DEFAULT_ACCOUNT_INITIAL_BALANCE)
+        - (MAX_PAYMENT_AMOUNT + TRANSFER_COST + TRANSFER_COST);
+    assert_eq!(
+        default_account_balance,
+        default_expected_balance,
+        "default account balance should reflect the transfer ({})",
+        U512::from(DEFAULT_ACCOUNT_INITIAL_BALANCE) - default_expected_balance
+    );
+
+    let account_1_balance = builder.get_purse_balance(account1.main_purse());
+    assert_eq!(
+        account_1_balance,
+        U512::zero(),
+        "account 1 balance should have been completely consumed"
+    );
+
+    let account_2_balance = builder.get_purse_balance(account2.main_purse());
+    assert_eq!(
+        account_2_balance, *TRANSFER_AMOUNT,
+        "account 2 balance should have changed"
+    );
+}
```

Comment:
Could we have used the `operations::execute_finalized_block` that hides the whole complexity of creation of `ScratchGlobalState`, commit and flushing its effects into the underlying LMDB?

Reply:
`contract_runtime::operations::execute_finalized_block` exists in the node, and relies on an input parameter `FinalizedBlock` which exists in the node, and we are upstream. I agree that `execute_finalized_block` should be lifted out of the node, however I don't think that should be done in this PR.